### ISSUE TITLE
Memory leak when call regions() in loop

### ISSUE THREAD

**Issue by nikkadim** [arrayfire/arrayfire#1943]
How to recreate:
```
int main(int argc, char** argv)
{

    try {
		af::setBackend(AF_BACKEND_CUDA);
		//af::setBackend(AF_BACKEND_CPU);
		//af::setBackend(AF_BACKEND_OPENCL);
		af::info();
		cudaFree(0);
		size_t free_byte ;
		size_t total_byte ;

		array L;
		array random;
		unsigned int IMGNUM = 20;

		for (int i=0; i<IMGNUM; i++) {
			
			random = i * randu(7000, 7000, u8);
			
			L = regions( random.as(b8), AF_CONNECTIVITY_4 );
			cudaMemGetInfo( &free_byte, &total_byte );
			fprintf(stdout,"Free:%d MB(%d MB)\n",free_byte/(1024*1024),total_byte/(1024*1024));
		}
    } catch (af::exception& e) {
        fprintf(stderr, "%s\n", e.what());
        throw;
    }
    return 0;
}
```

```
ArrayFire v3.5.1 (CUDA, 64-bit Windows, build 0a675e8)
Platform: CUDA Toolkit 8, Driver: CUDA Driver Version: 9000
[0] GeForce GTX 1070, 8192 MB, CUDA Compute 6.1
Free:6199 MB(8192 MB)
Free:5825 MB(8192 MB)
Free:5638 MB(8192 MB)
Free:5638 MB(8192 MB)
Free:5451 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5077 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4516 MB(8192 MB)
Free:4329 MB(8192 MB)
Free:4142 MB(8192 MB)
Free:4142 MB(8192 MB)
Free:4142 MB(8192 MB)
```

with deviceGC();
```
ArrayFire v3.5.1 (CUDA, 64-bit Windows, build 0a675e8)
Platform: CUDA Toolkit 8, Driver: CUDA Driver Version: 9000
[0] GeForce GTX 1070, 8192 MB, CUDA Compute 6.1
Free:6199 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
Free:5827 MB(8192 MB)
```

**Comment by pavanky** [comment]
If theres a leak it should happen with deviceGC as well. Can you show the outputs from af::deviceMemIno()?

**Comment by nikkadim** [comment]
@pavanky 
output with deviceMemInfo(...):
```
ArrayFire v3.5.1 (CUDA, 64-bit Windows, build 0a675e8)
Platform: CUDA Toolkit 8, Driver: CUDA Driver Version: 9000
[0] GeForce GTX 1070, 8192 MB, CUDA Compute 6.1
alloc_bytes:467 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:841 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1028 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1028 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1214 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1401 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1401 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1401 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1588 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1775 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1775 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1775 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1962 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1962 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:1962 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:2149 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:2336 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:2523 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:2523 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
alloc_bytes:2523 MB alloc_buffers:0 MB lock_bytes:233 MB lock_buffers:0 MB
```

**Comment by pavanky** [comment]
@nikkadim Can you try changing your code slightly to this:

```
random = (i * randu(7000, 7000, u8)).as(b8);
af::eval(random);
L = regions( random, AF_CONNECTIVITY_4 );
```

**Comment by nikkadim** [comment]
done:
```

ArrayFire v3.5.1 (CUDA, 64-bit Windows, build 0a675e8)
Platform: CUDA Toolkit 8, Driver: CUDA Driver Version: 9000
[0] GeForce GTX 1070, 8192 MB, CUDA Compute 6.1
Free:6199 MB(8192 MB)
Free:5825 MB(8192 MB)
Free:5638 MB(8192 MB)
Free:5638 MB(8192 MB)
Free:5451 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5077 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4516 MB(8192 MB)
Free:4329 MB(8192 MB)
Free:4142 MB(8192 MB)
Free:4142 MB(8192 MB)
Free:4142 MB(8192 MB)
```

**Comment by pavanky** [comment]
@nikkadim Ok I think I know what is happening. ArrayFire allocates new memory when the output size (rounded to the closest KB) is different. Since you are creating new image each time, the temporary buffers used internally are going to be different sizes (depending on the number of unique regions you have). If you created random outside the loop and used a copy of random everytime, you wouldnt see an increase in memory usage. 

Currently there's no way to fix this.. Unless you are OK with using a large memory step size.
To change the memory step size, you should add `af_set_mem_step_size(my_custom_step_size);` and tweak what `my_custom_step_size`. I'd assume if you set it to 1MB instead of 1KB, You shouldn't be seeing this.

**Comment by pavanky** [comment]
I also created this issue to fix this properly in arrayfire: https://github.com/arrayfire/arrayfire/issues/1946

**Comment by nikkadim** [comment]
@pavanky as you said:

**af_set_mem_step_size(1024*1024);**
```
ArrayFire v3.5.1 (CUDA, 64-bit Windows, build 0a675e8)
Platform: CUDA Toolkit 8, Driver: CUDA Driver Version: 9000
[0] GeForce GTX 1070, 8192 MB, CUDA Compute 6.1
Free:6198 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
Free:5824 MB(8192 MB)
```

**af_set_mem_step_size(1024);**
```
ArrayFire v3.5.1 (CUDA, 64-bit Windows, build 0a675e8)
Platform: CUDA Toolkit 8, Driver: CUDA Driver Version: 9000
[0] GeForce GTX 1070, 8192 MB, CUDA Compute 6.1
Free:6199 MB(8192 MB)
Free:5825 MB(8192 MB)
Free:5638 MB(8192 MB)
Free:5638 MB(8192 MB)
Free:5451 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5264 MB(8192 MB)
Free:5077 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4890 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4703 MB(8192 MB)
Free:4516 MB(8192 MB)
Free:4329 MB(8192 MB)
Free:4142 MB(8192 MB)
Free:4142 MB(8192 MB)
Free:4142 MB(8192 MB)
```

Thanks!

**Comment by pavanky** [comment]
@nikkadim Closing this issue. I'll fix this more correctly in #1946
